augur/tree.py

```diff
@@ -232,8 +232,23 @@ def random_string(n):
             tmp_line = line
             if line.startswith(">"):
                 num_seqs += 1
-                for c,v in escape_dict.items():
-                    tmp_line = tmp_line.replace(c,v)
+                pos = 1
+                for i in range(1, len(line)):
+                    l = line[i]
+                    #check if l is a character that IQ-tree changes
+                    # See IQtree function renameBool https://github.com/iqtree/iqtree2/blob/3bbc304263cb2f85574a9163e8f2e5c5b597a147/utils/tools.cpp#L585
+                    if l.isalnum()==False and l !='_' and l != '-' and l != '.' and l.isspace()== False:
+                        if l not in escape_dict: #characters outside of escape dictionary might not be properly handled in treetime 
+                            print("WARNING: Potentially offending character: \'%s\' detected in taxon name: %s" 
+                                "We recommend replacing offending characters with '_' in the alignment file to avoid issues downstream."
+                                % (l, format(line))
+                            )
+                            escape_dict[l] = f'_{prefix}-{random_string(20)}_'
+                            reverse_escape_dict[escape_dict[l]] = l
+                        tmp_line = tmp_line[:pos] + escape_dict[l] + line[(i+1):]  # replace offending characters
+                        pos = pos + len(escape_dict[l])
+                    else:
+                        pos += 1
```

Comment:
This sort of position tracking and iterative mutation (of `tmp_line`) can get confusing and hard to follow quick, esp. when coming back to it later. I think a functional approach, e.g. with `re.sub()` with a replacement function (or some similar approach), might be more clear. I've sketched out what that looks like: https://gist.github.com/tsibley/8b55c1b721b382cab4c6276442528778

Reply:
Thank you! I have incorporated your suggestions :-)